doc/design/docker-cp.md

```diff
@@ -0,0 +1,146 @@
+
+4. The container is *off*, an we want to copy data *to* it.
+
+According to the docker remote api the target for a copy must be in `identity (no compression), gzip, bzip2, xz`. As the docker cli handles packaging a *to* request we can expect on of these formats. We must package our *from* response in one of these formats as well. 
+
+ _NOTE_: docker has discussed copying between containers in the past. This will be yet another set of 4 possibilities and endpoints. Something that we can add to this plan later if it is needed in the future. 
+
+### Personality to Portlayer Communication
+
+Since all 4 of these must be supported we will require both distinctions for when the container is off and which endpoint to call to distinguish between *to* and *from* operations. 
+
+The proposed solution is to have two endpoints designed for the portlayer api swagger server. Both endpoints will exist along the same request path as two different verbs. The request path should be as such : 
+
+Ideally we would like a GRPC endpoint rather than a RESTful endpoint. @sflxn has already built out this kind of behavior and we will likely aim for this implementation rather than passing streams through a restful approach between the personality and the portlayerl.
+
+
+```
+/container/filesystem?ID=
+```
+
+For a copy *to* the filesystem of a container we will do a `PUT` operation
+
+```
+/contrainer/filesystem?ID= PUT
+
+Request Model : 
+
+* path - container filesystem target for the copy operation
+
+* ID - this is the target container
+
+* overwriteDir - flag for whether or not we are directed to overwrite existing directories 
+
+Response Model :
+
+* Error - error response from the portlayer
+
+```
+
+For a copy *from* the filesystem of a container we will do a `GET` operation. 
+
+```
+/container/filesystem?ID=
+
+
+Request Model : 
+
+* ID - this is the target container
+
+* path - target path for the copy operation
+
+Response Model : 
+
+* Error - error response from the portlayer
+
+```
+
+### Personality design
+
+The personality has two endpoints that are called depending on the to/from situation above. The portlayer will divide the behavior of to/from as two separate endpoints. These endpoints will behave in an `Export/Import` behavior. There will be several checks during this process. 
+
+Error codes expected by docker : 
+
+```
+200 
+
+400
+
+403 - in the event of a readonly filesystem
+
+404
+
+409
+
+500
+
+```
+
+Docker does not do much in the way of string checking as far as the path is concerned. This is likely not a big concern for the personality as the actual copy commmand will fail if a bad path string is provided. 
+
+The personality code should be fairly straight forward, the goal is to call the portlayer for the copy while passing the tar stream either to the portlayer or from it. The best way to do this is to hook the writiers together using a PipeReader and then shepherding the data in the direction it needs to go. 
```

Comment:
I think all you need to say here is that we can use the same streaming mechanism used for attach, exec, and logs.